File: benchmark/01_prime_sieve/sieve.go
// Benchmark 01: Prime Sieve (Eratosthenes)
// Category: Integer Arithmetic
// Tests: Array operations, basic arithmetic, memory access

package main

import (
	"fmt"
	"math"
	"time"
)

const LIMIT = 10000000 // 10 million

func main() {
	start := time.Now()

	// Allocate sieve array
	isPrime := make([]bool, LIMIT+1)
	for i := 2; i <= LIMIT; i++ {
		isPrime[i] = true
	}

	// Sieve of Eratosthenes
	sqrtLimit := int(math.Sqrt(float64(LIMIT)))
	for p := 2; p <= sqrtLimit; p++ {
		if isPrime[p] {
			for i := p * p; i <= LIMIT; i += p {
				isPrime[i] = false
			}
		}
	}

	// Count primes
	count := 0
	for i := 2; i <= LIMIT; i++ {
		if isPrime[i] {
			count++
		}
	}

	elapsed := time.Since(start)

	fmt.Printf("Primes up to %d: %d\n", LIMIT, count)
	fmt.Printf("Time: %.4f seconds\n", elapsed.Seconds())
}

File: benchmark/01_prime_sieve/sieve.rs
// Benchmark 01: Prime Sieve (Eratosthenes)
// Category: Integer Arithmetic
// Tests: Array operations, basic arithmetic, memory access

use std::time::Instant;

const LIMIT: usize = 10_000_000; // 10 million

fn main() {
    let start = Instant::now();

    // Allocate sieve array
    let mut is_prime = vec![true; LIMIT + 1];
    is_prime[0] = false;
    is_prime[1] = false;

    // Sieve of Eratosthenes
    let sqrt_limit = (LIMIT as f64).sqrt() as usize;
    for p in 2..=sqrt_limit {
        if is_prime[p] {
            for i in (p * p..=LIMIT).step_by(p) {
                is_prime[i] = false;
            }
        }
    }

    // Count primes
    let count = is_prime.iter().filter(|&&x| x).count();

    let elapsed = start.elapsed();

    println!("Primes up to {}: {}", LIMIT, count);
    println!("Time: {:.4} seconds", elapsed.as_secs_f64());
}
